src/realtime-callbacks.js

```diff
@@ -56,6 +56,80 @@ module.exports.setNow = function(f) {
 };
 let _now = Date.now;
 
+// Internal hash of active debouncers
+const _debouncers = {};
+
+class Debouncer {
+
+    constructor(jobName) {
+        this.jobName = jobName;
+    }
+
+    complete() {
+        delete _debouncers[this.jobName];
+        this.func();
+    }
+
+    start(func, waitMs) {
+        if (func !== undefined) this.func = func;
+        if (waitMs !== undefined) this.waitMs = waitMs;
+        this.timeoutId = setTimeout(this.complete.bind(this), this.waitMs);
+    }
+
+    stop() {
+        if (this.timeoutId) {
+            clearTimeout(this.timeoutId);
+            delete this.timeoutId;
+        }
+    }
+}
+
+/**
+ * Implementation of debounce a which will call the callback after
+ * a waitMs period if it is not debounced within that period.
+ *
+ * After a jobName has been defined it may be called without
+ * {func} or {waitMs} arguments to debounce with initial settings.
+ * New {func} or {waitMs} params will over-write ones previously
+ * set in the debouncer.
+ * If called without a jobName a random one is selected.
+ *
+ * @param {function} func   callback to be called after a delay
+ * @param {Number} waitMs  number of milliseconds to delay by
+ *
+ * @return {String} a jobName to reference the debouncer.
+ */
+
+module.exports.debounce = function(jobName, func, waitMs) {
+    if (jobName === undefined) {
+        while(_debouncers[jobName = Math.random().toString(16).slice(2, 10)]);
```

Comment:
Aaah, right I see now. This line feels a bit too clever, you're:
 - Generating the random ID.
 - Assigning it to `jobName`.
 - Keying into `_debouncers` to check if there's a collision.
 - If there is a collision, rinse and repeat else continue to the next line.

...all on one line. Please can you break this up onto multiple lines which are more self-explanatory, and even add a comment to say what you're doing e.g:

```js
// make sure the job name doesn't collide with an existing debouncer
while (!jobName || _debouncers[jobName]) {
    jobName = Math.random().toString(16).slice(2, 10);
}
```